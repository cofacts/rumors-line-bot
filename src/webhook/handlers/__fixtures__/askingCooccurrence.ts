import {
  ListArticlesInInitStateQuery,
  ListArticlesInProcessMediaQuery,
} from 'typegen/graphql';

export const textMessage = {
  id: 'text-message',
  text: 'This is a text message',
  articleType: 'TEXT',
  attachmentUrl: null,
  replyCount: 0,
} as const satisfies NonNullable<
  ListArticlesInInitStateQuery['ListArticles']
>['edges'][0]['node'];

export const imageMessage = {
  id: 'img-message',
  articleType: 'TEXT',
  attachmentUrl: 'https://example.com/image.jpg',
  replyCount: 0,
} as const satisfies NonNullable<
  ListArticlesInProcessMediaQuery['ListArticles']
>['edges'][0]['node'];
