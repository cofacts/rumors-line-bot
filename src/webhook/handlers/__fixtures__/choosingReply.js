export const oneReply = {
  data: {
    GetReply: {
      type: 'NOT_ARTICLE',
      text: '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
      createdAt: '2018-01-09T05:52:12.658Z',
    },
    GetArticle: {
      replyCount: 1,
    },
  },
};

export const oneReply2 = {
  data: {
    GetReply: {
      type: 'RUMOR',
      text: 'The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply The Only Reply ',
      reference: '',
      createdAt: '2018-01-09T05:52:12.658Z',
    },
    GetArticle: {
      replyCount: 1,
    },
  },
};

export const multipleReplies = {
  data: {
    GetReply: {
      type: 'NOT_ARTICLE',
      text: '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
      createdAt: '2018-01-09T05:52:12.658Z',
    },
    GetArticle: {
      replyCount: 4,
    },
  },
};
