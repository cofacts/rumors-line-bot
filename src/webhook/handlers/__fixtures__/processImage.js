export const oneImageArticle = {
  data: {
    ListArticles: {
      edges: [
        {
          score: 1.87,
          node: {
            articleType: 'IMAGE',
            attachmentHash: 'ffff8000',
            attachmentUrl: 'http://foo/image.jpeg',
            id: 'image-article-1',
          },
        },
      ],
    },
  },
};

export const oneIdenticalImageArticle = {
  data: {
    ListArticles: {
      edges: [
        {
          score: 2,
          node: {
            articleType: 'IMAGE',
            attachmentHash: 'ffff8000',
            attachmentUrl: 'http://foo/image.jpeg',
            id: 'image-article-1',
          },
        },
      ],
    },
  },
};

export const twelveImageArticles = {
  data: {
    ListArticles: {
      edges: Array.from(Array(6)).reduce(
        arr =>
          arr.concat([
            {
              score: 2,
              node: {
                articleType: 'IMAGE',
                attachmentHash: 'ffff8000',
                attachmentUrl: 'http://foo/image.jpeg',
                id: 'image-article-1',
              },
            },
            {
              score: 1.876656,
              node: {
                articleType: 'IMAGE',
                attachmentHash: 'ffff8001',
                attachmentUrl: 'http://foo/image2.jpeg',
                id: 'image-article-2',
              },
            },
          ]),
        []
      ),
    },
  },
};

export const notFound = {
  data: {
    ListArticles: {
      edges: [],
    },
  },
};
