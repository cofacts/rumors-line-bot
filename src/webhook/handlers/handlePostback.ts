import { t } from 'ttag';
import choosingArticle from './choosingArticle';
import choosingReply from './choosingReply';
import askingArticleSubmissionConsent from './askingArticleSubmissionConsent';
import askingArticleSource from './askingArticleSource';
import defaultState from './defaultState';
import askingCooccurence from './askingCooccurrence';
import { ManipulationError } from './utils';
import tutorial from './tutorial';
import {
  ChatbotPostbackHandlerParams,
  Result,
  Context,
  PostbackActionData,
} from 'src/types/chatbotState';

/**
 * Given input event and context, outputs the new context and the reply to emit.
 *
 * @param data The current context of the bot
 * @param postbackData The input postback data extracted from event
 * @param userId LINE user ID that does the input
 */
export default async function handlePostback(
  context: Context,
  postbackData: PostbackActionData<unknown>,
  userId: string
) {
  const params: ChatbotPostbackHandlerParams = {
    context,
    postbackData,
    userId,
  };

  let result: Result;

  // Sets data and replies
  //
  try {
    switch (params.postbackData.state) {
      case 'CHOOSING_ARTICLE': {
        result = await choosingArticle(params);
        break;
      }
      case 'CHOOSING_REPLY': {
        result = await choosingReply(params);
        break;
      }
      case 'TUTORIAL': {
        result = tutorial(params);
        break;
      }
      case 'ASKING_ARTICLE_SOURCE': {
        result = await askingArticleSource(params);
        break;
      }
      case 'ASKING_ARTICLE_SUBMISSION_CONSENT': {
        result = await askingArticleSubmissionConsent(params);
        break;
      }
      case 'ASKING_COOCCURRENCE': {
        result = await askingCooccurence(params);
        break;
      }
      case 'CONTINUE': {
        // Do nothing; pass context (updated by singleUserHandler) as-is and reply nothing.
        result = {
          context,
          replies: [],
        };
        break;
      }
      default: {
        result = defaultState(params);
        break;
      }
    }
  } catch (e) {
    if (e instanceof ManipulationError) {
      result = {
        context,
        replies: [
          {
            type: 'flex',
            altText: e.toString(),
            contents: {
              type: 'bubble',
              header: {
                type: 'box',
                layout: 'vertical',
                contents: [
                  {
                    type: 'text',
                    text: `⚠️ ${t`Wrong usage`}`,
                    color: '#ffb600',
                    weight: 'bold',
                  },
                ],
              },
              body: {
                type: 'box',
                layout: 'vertical',
                contents: [
                  {
                    type: 'text',
                    text: e.message,
                    wrap: true,
                  },
                ],
              },
              styles: {
                body: {
                  separator: true,
                },
              },
            },
          },
        ],
      };
    } else {
      throw e;
    }
  }

  return result;
}
