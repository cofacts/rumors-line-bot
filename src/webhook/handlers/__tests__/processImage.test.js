jest.mock('src/lib/gql');
jest.mock('src/lib/ga');
jest.mock('src/lib/detectDialogflowIntent');
jest.mock('src/webhook/handlePostback', () =>
  jest.fn(() => '__HANDLE_POSTBACK_RESULT__')
);

import MockDate from 'mockdate';
import processImage from '../processImage';
import * as apiResult from '../__fixtures__/processImage';
import gql from 'src/lib/gql';
import ga from 'src/lib/ga';
import handlePostback from 'src/webhook/handlePostback';

beforeEach(() => {
  ga.clearAllMocks();
  gql.__reset();
  handlePostback.mockClear();
});

it('one identical article found', async () => {
  gql.__push(apiResult.oneIdenticalImageArticle);

  const data = {
    sessionId: 1497994017447,
  };
  const event = {
    type: 'message',
    timestamp: 1497994016356,
    messageId: '6270464463537',
    message: {
      type: 'image',
      id: '6270464463537',
    },
  };
  const userId = 'Uc76d8ae9ccd1ada4f06c4e1515d46466';
  MockDate.set('2020-01-01');
  expect(await processImage(data, event, userId)).toMatchSnapshot();
  MockDate.reset();
  expect(gql.__finished()).toBe(true);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "MessageType",
          "ec": "UserInput",
          "el": "image",
        },
      ],
      Array [
        Object {
          "ea": "ArticleSearch",
          "ec": "UserInput",
          "el": "ArticleFound",
        },
      ],
      Array [
        Object {
          "ea": "Search",
          "ec": "Article",
          "el": "image-article-1",
          "ni": true,
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});

it('one article found (not identical)', async () => {
  gql.__push(apiResult.oneImageArticle);

  const data = {
    sessionId: 1497994017447,
  };
  const event = {
    type: 'message',
    timestamp: 1497994016356,
    messageId: '6270464463537',
    message: {
      type: 'image',
      id: '6270464463537',
    },
  };
  const userId = 'Uc76d8ae9ccd1ada4f06c4e1515d46466';
  MockDate.set('2020-01-01');
  expect(await processImage(data, event, userId)).toMatchSnapshot();
  MockDate.reset();
  expect(gql.__finished()).toBe(true);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "MessageType",
          "ec": "UserInput",
          "el": "image",
        },
      ],
      Array [
        Object {
          "ea": "ArticleSearch",
          "ec": "UserInput",
          "el": "ArticleFound",
        },
      ],
      Array [
        Object {
          "ea": "Search",
          "ec": "Article",
          "el": "image-article-1",
          "ni": true,
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});

it('twelve articles found', async () => {
  gql.__push(apiResult.twelveImageArticles);

  const data = {
    sessionId: 1497994017447,
  };
  const event = {
    type: 'message',
    timestamp: 1497994016356,
    messageId: '6530038889933',
    message: {
      type: 'image',
      id: '6530038889933',
    },
  };
  const userId = 'Uc76d8ae9ccd1ada4f06c4e1515d46466';

  MockDate.set('2020-01-01');
  const result = await processImage(data, event, userId);
  MockDate.reset();
  expect(result).toMatchSnapshot();
  expect(gql.__finished()).toBe(true);
  expect(result.replies.length).toBeLessThanOrEqual(5); // Reply message API limit
  const carousel = result.replies.find(({ type }) => type === 'flex').contents;
  expect(carousel.type).toBe('carousel');
  expect(carousel.contents.length).toBeLessThanOrEqual(10); // Flex message carousel 10 bubble limit
  expect(JSON.stringify(carousel).length).toBeLessThan(50 * 1000); // Flex message carousel 50K limit
});

it('should handle image not found', async () => {
  gql.__push(apiResult.notFound);
  const data = {
    sessionId: 1497994017447,
  };
  const event = {
    type: 'message',
    timestamp: 1497994016356,
    messageId: '6530038889933',
    message: {
      type: 'image',
      id: '6530038889933',
    },
  };
  const userId = 'Uc76d8ae9ccd1ada4f06c4e1515d46466';
  MockDate.set('2020-01-01');
  expect(await processImage(data, event, userId)).toMatchSnapshot();
  MockDate.reset();
  expect(gql.__finished()).toBe(true);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "MessageType",
          "ec": "UserInput",
          "el": "image",
        },
      ],
      Array [
        Object {
          "ea": "ArticleSearch",
          "ec": "UserInput",
          "el": "ArticleNotFound",
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});
