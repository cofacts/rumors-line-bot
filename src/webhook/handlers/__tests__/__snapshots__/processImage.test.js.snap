// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one article found (not identical) 1`] = `
Object {
  "context": Object {
    "data": Object {
      "messageId": "6270464463537",
      "messageType": "image",
      "searchedText": "",
      "sessionId": 1577836800000,
    },
  },
  "replies": Array [
    Object {
      "text": "🔍 There are some messages that looks similar to the one you have sent to me.",
      "type": "text",
    },
    Object {
      "text": "Internet rumors are often mutated and shared.
Please choose the version that looks the most similar👇",
      "type": "text",
    },
    Object {
      "altText": "Please choose the most similar message from the list.",
      "contents": Object {
        "contents": Array [
          Object {
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose “No.1”",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😄",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "No.1, Looks 87% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "gravity": "top",
                  "maxLines": 5,
                  "text": "None of these messages matches mine :(",
                  "type": "text",
                  "weight": "regular",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "margin": "none",
              "spacing": "none",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"__NO_ARTICLE_FOUND__\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "None of these messages matches mine :(",
                    "label": "Tell us more",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "margin": "none",
                  "size": "sm",
                  "text": "😶",
                  "type": "text",
                  "weight": "bold",
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "type": "box",
            },
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
}
`;

exports[`one identical article found and choose for user 1`] = `
Object {
  "data": Object {
    "messageId": "6270464463537",
    "messageType": "image",
    "searchedText": "",
    "selectedArticleId": "image-article-1",
    "selectedArticleText": "",
    "sessionId": 1577836800000,
  },
  "event": Object {
    "input": "image-article-1",
    "type": "server_choose",
  },
  "replies": Array [
    Object {
      "text": "👨‍👩‍👧‍👦  Volunteer editors have published several replies to this message.

2 of them say it ❌ contains misinformation.",
      "type": "text",
    },
    Object {
      "text": "Let's pick one 👇",
      "type": "text",
    },
    Object {
      "altText": "Please take a look at the following replies.",
      "contents": Object {
        "contents": Array [
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "align": "start",
                  "margin": "md",
                  "maxLines": 10,
                  "text": "查證這則謠言的過程中，約翰走鹿也發現其實網路早就有許多解答，就讓約翰走鹿來借花獻佛一下囉。多家媒體都有針對這則「吃生魚片會有海獸胃腺蟲附著胃壁」的謠言請教醫師，受訪的醫師們皆表示，這則謠言太誇大，海獸胃腺蟲沒那麼神通廣大。
醫師們表示，台灣的衛生條件很進步，吃生魚片被寄生蟲寄生的案例很少，大部份都是生魚片不乾淨，感染沙門氏菌、大腸桿菌的案例，此外，寄生蟲也不是什麼難治之症，服用一些抗寄生蟲藥就可以殺死胃裡的這些蟲子，除非這些蟲子跑到身體其他部位那才比較難處理，但那也是要大量地吃生魚片（一週兩次以上），中獎機率才會比較高。",
                  "type": "text",
                  "wrap": true,
                },
                Object {
                  "type": "filler",
                },
                Object {
                  "margin": "md",
                  "type": "separator",
                },
                Object {
                  "contents": Array [
                    Object {
                      "size": "xs",
                      "text": "👍 14 users consider this helpful
😕 2 users consider this not useful",
                      "type": "text",
                      "wrap": true,
                    },
                  ],
                  "layout": "horizontal",
                  "margin": "md",
                  "spacing": "none",
                  "type": "box",
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"AVygFA0RyCdS-nWhuaXY\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_REPLY\\"}",
                    "displayText": "I choose “查證這則謠言的過程中，約翰走鹿也發現其實網路早⋯⋯”",
                    "label": "👀 Take a look",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "💬",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "Someone thinks it contains misinformation",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "align": "start",
                  "margin": "md",
                  "maxLines": 10,
                  "text": "過去台灣只有一個海獸胃腺蟲鑽入胃壁的病例，而且僅有一隻蟲。而經查詢，台北馬偕醫院也沒有名為林清風的醫師。",
                  "type": "text",
                  "wrap": true,
                },
                Object {
                  "type": "filler",
                },
                Object {
                  "margin": "md",
                  "type": "separator",
                },
                Object {
                  "contents": Array [
                    Object {
                      "size": "xs",
                      "text": "👍 24 users consider this helpful",
                      "type": "text",
                      "wrap": true,
                    },
                  ],
                  "layout": "horizontal",
                  "margin": "md",
                  "spacing": "none",
                  "type": "box",
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"AVy6LkWIyCdS-nWhuaqu\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_REPLY\\"}",
                    "displayText": "I choose “過去台灣只有一個海獸胃腺蟲鑽入胃壁的病例，而且⋯⋯”",
                    "label": "👀 Take a look",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "💬",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "Someone thinks it contains misinformation",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`should handle image not found 1`] = `
Object {
  "context": Object {
    "data": Object {
      "messageId": "6530038889933",
      "messageType": "image",
      "searchedText": "",
      "sessionId": 1577836800000,
    },
  },
  "replies": Array [
    Object {
      "altText": "Unfortunately, I currently don’t recognize this message, but I would still like to help.
Do you want someone to fact-check this message?",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "Unfortunately, I currently don’t recognize this message, but I would still like to help.
Do you want someone to fact-check this message?",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "Be the first to report the message",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "text": "Currently we don’t have this message in our database. If you think it is most likely a rumor, ",
                  "type": "span",
                },
                Object {
                  "color": "#ffb600",
                  "text": "press “🆕 Report to database” to make this message public on Cofacts database ",
                  "type": "span",
                  "weight": "bold",
                },
                Object {
                  "text": "and have volunteers fact-check it. This way you can help the people who receive the same message in the future.",
                  "type": "span",
                },
              ],
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "paddingAll": "lg",
          "spacing": "md",
          "type": "box",
        },
        "footer": Object {
          "contents": Array [
            Object {
              "action": Object {
                "data": "{\\"input\\":\\"__POSTBACK_YES__\\",\\"sessionId\\":1577836800000,\\"state\\":\\"ASKING_ARTICLE_SUBMISSION_CONSENT\\"}",
                "displayText": "🆕 Report to database",
                "label": "🆕 Report to database",
                "type": "postback",
              },
              "color": "#ffb600",
              "style": "primary",
              "type": "button",
            },
            Object {
              "action": Object {
                "data": "{\\"input\\":\\"__POSTBACK_NO__\\",\\"sessionId\\":1577836800000,\\"state\\":\\"ASKING_ARTICLE_SUBMISSION_CONSENT\\"}",
                "displayText": "Don’t report",
                "label": "Don’t report",
                "type": "postback",
              },
              "color": "#333333",
              "style": "primary",
              "type": "button",
            },
          ],
          "layout": "vertical",
          "spacing": "sm",
          "type": "box",
        },
        "styles": Object {
          "body": Object {
            "separator": true,
          },
        },
        "type": "bubble",
      },
      "type": "flex",
    },
  ],
}
`;

exports[`twelve articles found 1`] = `
Object {
  "context": Object {
    "data": Object {
      "messageId": "6530038889933",
      "messageType": "image",
      "searchedText": "",
      "sessionId": 1577836800000,
    },
  },
  "replies": Array [
    Object {
      "text": "🔍 There are some messages that looks similar to the one you have sent to me.",
      "type": "text",
    },
    Object {
      "text": "Internet rumors are often mutated and shared.
Please choose the version that looks the most similar👇",
      "type": "text",
    },
    Object {
      "altText": "Please choose the most similar message from the list.",
      "contents": Object {
        "contents": Array [
          Object {
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose “No.1”",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😄",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "No.1, Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-2\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose “No.2”",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😄",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "No.2, Looks 87% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image2.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose “No.3”",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😄",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "No.3, Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-2\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose “No.4”",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😄",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "No.4, Looks 87% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image2.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose “No.5”",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😄",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "No.5, Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-2\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose “No.6”",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😄",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "No.6, Looks 87% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image2.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose “No.7”",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😄",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "No.7, Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-2\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose “No.8”",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😄",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "No.8, Looks 87% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image2.jpeg",
            },
            "type": "bubble",
          },
          Object {
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":\\"image-article-1\\",\\"sessionId\\":1577836800000,\\"state\\":\\"CHOOSING_ARTICLE\\"}",
                    "displayText": "I choose “No.9”",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😄",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "No.9, Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "hero": Object {
              "size": "full",
              "type": "image",
              "url": "http://foo/image.jpeg",
            },
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
}
`;
